### PR TITLE
Hardening of readonly variables

### PR DESCRIPTION
**Pull request by hyenias**
Ksh currently restricts readonly scalar variables from having their values
directly changed via a value assignment. However, since ksh allows variable
attributes to be altered; this allows the variable's value to be indirectly
altered. For instance, if TMOUT=900 (for a 15 minute idle timeout) was set
to readonly; all that is needed to alter the value of TMOUT from 900 to 0 is
to issue 'typeset -R1 TMOUT' then perhaps following with a 'typeset -i TMOUT'
to turn off the shell's timeout value.

It appears that prior to Version JM 93u- 2010-06-29 (multishell ms_2010-07-01
branch), ksh did not allow TMOUT's attributes to be altered once it became
readonly.

src/cmd/ksh93/bltins/typeset.c: setall():
- Relocate readonly attribute check higher up the code as well as widen its
  application to issue an error message if the pre-existing name-pair has the
  readonly bit flag set.

src/cmd/ksh93/tests/variables.sh:
- KSH_VERSION is a nameref to .sh.version. As such, when a test ran and applied
  readonly to it ('typeset -n KSH_VERSION=.sh.version'); that readonly action
  applied to its named reference variable of .sh.version not to itself. As such,
  the subsequent readonly action upon .sh.version variable failed. Remove
  KSH_VERSION nameref from the test.

### CODE REVIEW

**Comment by hyenias** [comment]
I am still working on this but need some help/pointers.

Currently, the following is still being allowed as it appears it has something to do with indexed/associative arrays and just empty arrays.  I have tried everything I can think of under setall() but it appears stuff is happening elsewhere. I have tried other files as well with no luck. Need help. Can someone point me to the .c file(s) and maybe the function to go check into where ksh sets an array subscript?
1. arch/*/bin/ksh -c 'typeset -a arr=((a b c) 1); readonly arr; arr[0][1]=d; echo ${arr[0][1]}'
2. arch/*/bin/ksh -c 'arr=(alphas=(a b c);name=x); readonly arr.alphas; arr.alphas[1]=([b]=5); typeset -p arr arr.alphas'
3. arch/*/bin/ksh -c 'arr=(alphas=(a b c);name=x); readonly arr.alphas; arr.alphas[1]=(b); typeset -p arr arr.alphas'
4. arch/*/bin/ksh -c 'typeset -C arr=(typeset -r -a alphas=(a b c);name=x); arr.alphas[1]=(); typeset -p arr'

Also, if you attempt to unset the named reference KSH_VERSION; a memory fault occurs. I plan on using this to learn how to do some gdb tracing.
```ksh
$ unset -n KSH_VERSION
Memory fault
MacBook-Air:~$
```

**Comment by JohnoKing** [comment]
> Can someone point me to the .c file(s) and maybe the function to go check into where ksh sets an array subscript?

`nv_putsub` in `src/cmd/ksh93/array.c` might be relevant:
https://github.com/ksh93/ksh/blob/814b5c6890b897d0e7b2625a16a6513aa5d2789b/src/cmd/ksh93/sh/array.c#L1152-L1164

**Comment by hyenias** [comment]
@JohnoKing, thank you. I will have a look at it.

**Comment by McDutchie** [comment]
The current PR makes it an error to declare a variable readonly that is already readonly:
```
$ arch/*/bin/ksh -c 'readonly foo; readonly foo'
arch/darwin.i386-64/bin/ksh: readonly: foo: is read only
```
and also:
```
$ arch/*/bin/ksh -c 'readonly foo foo'
arch/darwin.i386-64/bin/ksh: readonly: foo: is read only
```
The second one is why the regression test needed changing. That change should not be necessary.

Almost no other shell does this. More importantly, this may break some existing ksh93 scripts, so we should avoid introducing this incompatibility.

**Comment by McDutchie** [comment]
A little debugging (diff against current PR):
```diff
--- a/src/cmd/ksh93/bltins/typeset.c
+++ b/src/cmd/ksh93/bltins/typeset.c
@@ -701,6 +701,7 @@ static int     setall(char **argv,register int flag,Dt_t *troot,struct tdata *tp
 			np = nv_open(name,troot,nvflags|((nvflags&NV_ASSIGN)?0:NV_ARRAY)|((iarray|(nvflags&(NV_REF|NV_NOADD)==NV_REF))?NV_FARRAY:0));
 			if(!np)
 				continue;
+error(ERROR_warn(0),"[DEBUG] flag: %d, 0%o, 0x%X",flag,flag,flag);
 			if(np->nvflag&NV_RDONLY && !tp->pflag)
 				errormsg(SH_DICT,ERROR_exit(1),e_readonly,nv_name(np));
 			if(nv_isnull(np) && !nv_isarray(np) && nv_isattr(np,NV_NOFREE))
```
After that:
```
$ arch/*/bin/ksh -c 'readonly foo; readonly foo'
arch/darwin.i386-64/bin/ksh: readonly: warning: [DEBUG] flag: 513, 01001, 0x201
arch/darwin.i386-64/bin/ksh: readonly: warning: [DEBUG] flag: 513, 01001, 0x201
arch/darwin.i386-64/bin/ksh: readonly: foo: is read only
$ arch/*/bin/ksh -c 'readonly foo foo'          
arch/darwin.i386-64/bin/ksh: readonly: warning: [DEBUG] flag: 513, 01001, 0x201
arch/darwin.i386-64/bin/ksh: readonly: warning: [DEBUG] flag: 1, 01, 0x1
arch/darwin.i386-64/bin/ksh: readonly: foo: is read only
```
So, if `foo` is readonly and passed as the first argument, `flag` is `0x201`. Looking at [`src/cmd/ksh93/include/nval.h`](https://github.com/ksh93/ksh/blob/f8de1f111dc0d87e0563ca055b639f39e0f66380/src/cmd/ksh93/include/nval.h#L137-L203) shows that this is `NV_RDONLY` or-ed with `NV_NOFREE`. Whereas if it is passed as the second argument, `flag` is simply `0x1` which is `NV_RDONLY`.

So, I think this problem can be fixed by simply whitelisting those values. This also removes the need to make any change to the regression test. I'll push a commit, please test it.

**Comment by hyenias** [comment]
I have been working on the other readonly changes involving subscripting so this PR is not ready. By allowing an existing variable to have the readonly attribute reapplied will make things more difficult to distinguish when to not allow a change.

I will pick this up tomorrow and test.

**Comment by McDutchie** [comment]
OK, sorry, didn't mean to interfere.

If it helps, here is a debug diff that prints out the file and line number of every `nv_putsub()` call as it happens, it may help you find the one(s) that need checks.

```diff
--- a/src/cmd/ksh93/include/nval.h
+++ b/src/cmd/ksh93/include/nval.h
@@ -250,7 +250,8 @@ extern void	*nv_associative(Namval_t*,const char*,int);
 extern int	nv_aindex(Namval_t*);
 extern int	nv_nextsub(Namval_t*);
 extern char	*nv_getsub(Namval_t*);
-extern Namval_t	*nv_putsub(Namval_t*, char*, long);
+extern Namval_t	*_real_nv_putsub(char*,int,Namval_t*, char*, long);
+#define nv_putsub(n,c,l) _real_nv_putsub(__FILE__,__LINE__,n,c,l)
 extern Namval_t	*nv_opensub(Namval_t*);
 
 /* name-value pair function prototypes */
diff --git a/src/cmd/ksh93/sh/array.c b/src/cmd/ksh93/sh/array.c
index 6decffa0..a827b365 100644
--- a/src/cmd/ksh93/sh/array.c
+++ b/src/cmd/ksh93/sh/array.c
@@ -1160,8 +1160,9 @@ int nv_nextsub(Namval_t *np)
  *   ARRAY_ADD is specified and there is no value or sets all
  * the elements up to the number specified if ARRAY_ADD is not specified
  */
-Namval_t *nv_putsub(Namval_t *np,register char *sp,register long mode)
+Namval_t *_real_nv_putsub(char *file,int line,Namval_t *np,register char *sp,register long mode)
 {
+error(ERROR_warn(0),"[DEBUG] nv_putsub() called from %s, line %d",file,line);
 	register struct index_array *ap = (struct index_array*)nv_arrayptr(np);
 	register int size = (mode&ARRAY_MASK);
 #if SHOPT_FIXEDARRAY
```

**Comment by hyenias** [comment]
Oh, heavens no. You have not interfered. I directly asked for help. Much appreciated for all the clues and even a neat little diff that just blew my mind. I spent many hours today and more last week going through name.c and array.c trying to follow the code to find the right spot that actually passes an intact NV_RDONLY flag to check against.

**Comment by hyenias** [comment]
I am happy to report that after 1 hour of attempting to alter TMOUT with the present PR using:

https://github.com/ksh93/ksh/blob/3fdd44827693c73af3f6e5c59cfea573fa5c66b7/src/cmd/ksh93/bltins/typeset.c#L704-L705

All attempts have failed. This concludes the hardening of the readonly scalar variables.

Continuing to work on hardening arrays.

**Comment by hyenias** [comment]
Please try this current PR commit as it now enables readonly for arrays subscripts. It seems to work with all of my hand testing. @McDutchie, the debugging code was helpful.

I will now start working on some readonly array tests as it does not appear that there are too many.

**Comment by hyenias** [comment]
Need help on fixing the new readonly.sh file as the number of tests stay at 2 and the number of errors do not increment. I do not understand how to get them to increment.

**Comment by hyenias** [comment]
Amazing what a little sleep will do for you. I found the missing code I overlooked that reports back the number of errors.

**Comment by hyenias** [comment]
Ready to have this PR reviewed then maybe merged unless I need to go back and adjust the if statements with `errormsg` for #248 and add in the `UNREACHABLE();` statements.

File: src/cmd/ksh93/tests/readonly.sh [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+#             Copyright (c) 2021 Contributors to ksh 93u+m             #
+#                    <https://github.com/ksh93/ksh>                    #
+#                      and is licensed under the                       #
+#                 Eclipse Public License, Version 1.0                  #
+#                                                                      #
+#                A copy of the License is available at                 #
+#          http://www.eclipse.org/org/documents/epl-v10.html           #
+#         (with md5 checksum b35adb5213ca9657e911e9befb180842)         #
+#                                                                      #
+#                  Martijn Dekker <martijn@inlv.org>                   #
```
Comment (McDutchie) [inline src/cmd/ksh93/tests/readonly.sh]: Wrong author name. :)

File: src/cmd/ksh93/tests/readonly.sh [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+		exp='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+		chg='arr[a][13]=XIII'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+	)
+
+# Below tests cause a 'recursion too deep' error condition
```
Comment (McDutchie) [inline src/cmd/ksh93/tests/readonly.sh]: This makes no sense -- it's some kind of strange bug in 93u+ and 93u+m. There is no recursion here! I tried uncommenting those tests and running them on various ksh versions. The error does not occur when running the tests on 93v- or ksh2020.

Comment (hyenias) [reply src/cmd/ksh93/tests/readonly.sh]: I figured it was a bug with arrays, compound variables, and/or arithmetic operators (#168) as I can manually run them. Hopefully, once that and other array issues get resolved those tests can be enabled.

File: src/cmd/ksh93/tests/readonly.sh (hunk 1 above)
Comment (hyenias) [reply src/cmd/ksh93/tests/readonly.sh]: Correcting to myself aka hyenias.

File: src/cmd/ksh93/tests/readonly.sh (hunk 2 above)
Comment (McDutchie) [reply src/cmd/ksh93/tests/readonly.sh]: I did some debugging and looked at the differences between 93u+ and 93v- in streval.c which handles arithmetic expansion. The [DEBUG] output shows where the recursion level is spuriously increased, which is informative when running your tests. The [DEBUG] output on 93v- is identical, so nothing changed there, except that they simply changed `MAXLEVEL` from 9 to 1024 (!) in that version, which avoids the problem. Fixing this properly is for another commit, but for now I think you should increase `MAXLEVEL` to 1024 as in 93v- and enable the rest of the tests.
```diff
--- a/src/cmd/ksh93/sh/streval.c
+++ b/src/cmd/ksh93/sh/streval.c
@@ -47,7 +47,7 @@
 #   define SH_DICT	"libshell"
 #endif
 
-#define MAXLEVEL	9
+#define MAXLEVEL	1024
 #define SMALL_STACK	12
 
 /*
@@ -177,6 +177,7 @@ Sfdouble_t	arith_exec(Arith_t *ep)
 	node.nosub = 0;
 	node.ptr = 0;
 	node.eflag = 0;
+error(ERROR_warn(0),"[DEBUG] level==%d",level);
 	if(level++ >=MAXLEVEL)
 	{
 		arith_error(e_recursive,ep->expr,ep->emode);
```

Comment (hyenias) [reply src/cmd/ksh93/tests/readonly.sh]: oooOOOOooo. nice. I go test it now.

Comment (hyenias) [reply src/cmd/ksh93/tests/readonly.sh]: Your DEBUG output is GOLD! I added in a little echo for the row and its initial setup. It is the `-i'`s that directly cause the level to increase which start on rtests[25].

```diff
diff --git a/src/cmd/ksh93/sh/streval.c b/src/cmd/ksh93/sh/streval.c
index 9a4e552..d573a03 100644
--- a/src/cmd/ksh93/sh/streval.c
+++ b/src/cmd/ksh93/sh/streval.c
@@ -47,7 +47,7 @@
 #   define SH_DICT     "libshell"
 #endif

-#define MAXLEVEL       9
+#define MAXLEVEL       1024
 #define SMALL_STACK    12

 /*
@@ -177,6 +177,7 @@ Sfdouble_t  arith_exec(Arith_t *ep)
        node.nosub = 0;
        node.ptr = 0;
        node.eflag = 0;
+error(ERROR_warn(0),"[DEBUG] level==%d",level);
        if(level++ >=MAXLEVEL)
        {
                arith_error(e_recursive,ep->expr,ep->emode);
diff --git a/src/cmd/ksh93/tests/readonly.sh b/src/cmd/ksh93/tests/readonly.sh
index e7e9756..d1d43e7 100644
--- a/src/cmd/ksh93/tests/readonly.sh
+++ b/src/cmd/ksh93/tests/readonly.sh
@@ -266,12 +266,12 @@ rtests=(
        )

 # Below tests cause a 'recursion too deep' error condition
-#      (
-#              ini='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
-#              chg='(( arr[b] += 5 ))'
-#              res='typeset -p arr'
-#              exp='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
-#      )
+       (
+               ini='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
+               chg='(( arr[b] += 5 ))'
+               res='typeset -p arr'
+               exp='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
+       )
 #      (
 #              ini='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
 #              chg='(( arr[b] -= 5 ))'
@@ -321,6 +321,7 @@ typeset -i i
 n=${#rtests[@]}
 for ((i=0; i<$n; i++))
 do
+       echo $i: "${rtests[$i].ini}"
        got=$(
                trap "${rtests[$i].res}" EXIT
                eval "${rtests[$i].ini}"
```
```
24: typeset -ra arr=((a b c) 1)
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==0
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==0
25: typeset -ra -i arr=((10 11 12) 3 4)
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==1
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==1
26: typeset -ra -i arr=((10 11 12) 3 4)
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==2
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==2
```

Comment (hyenias) [reply src/cmd/ksh93/tests/readonly.sh]: All done.

**Comment by McDutchie** [comment]
LGTM. Great work.

**Pull request merged**